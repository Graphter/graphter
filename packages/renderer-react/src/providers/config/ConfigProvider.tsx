import React, {createContext, useContext} from 'react';
import { NodeConfig, PathSegment } from "@graphter/core";

interface ConfigProviderProps {
  configs: Array<NodeConfig>
  children: any
}

const Context = createContext<Map<string | number, NodeConfig> | null>(null);

let configMap: Map<string | number, NodeConfig>

export function getConfig(id: string | number): NodeConfig {
  if(!configMap) throw new Error(`Couldn't find the API service registry. Make sure you've declared a <ServiceProvider /> and passed it a valid service.`)
  const service = configMap.get(id)
  if(!service) throw new Error(`Missing '${id}' config`)
  return service
}

export function useConfig(id: string | number): NodeConfig {
  const configMap = useContext(Context)
  if(!configMap) throw new Error(`Couldn't find the config registry. Make sure you've declared a <ConfigProvider /> and passed it valid config.`)
  const config = configMap.get(id)
  if(!config) throw new Error(`Missing '${id}' config`)
  return config
}

export default function ConfigProvider({ configs, children }: ConfigProviderProps){
  configMap = configs.reduce((a, c) => {
    if(Array.isArray(c)) c.forEach(registration => a.set(registration.id, registration.config))
    else a.set(c.id, c)
    return a
  }, new Map<string, NodeConfig>())

  return (
    <Context.Provider value={configMap}>
      {children}
    </Context.Provider>
  );
}