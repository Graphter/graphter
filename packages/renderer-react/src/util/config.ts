import { NodeConfig, PathSegment } from "@graphter/core";


const getAt = (config: NodeConfig | undefined, path?: Array<PathSegment> | null) => {
  const targetConfig = path?.reduce((config: NodeConfig | undefined, pathSegment: PathSegment) => {
    return config?.children?.find(childConfig => childConfig.id === pathSegment)
  }, config)
  if(typeof targetConfig === 'undefined'){
    throw new Error(`Could not find config at ${path?.join('/')} within: ${JSON.stringify(config)}.`)
  }
  return targetConfig
}

export const configUtils = {
  getAt
}