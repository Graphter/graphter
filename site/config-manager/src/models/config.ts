import { AllValidationExecutionStages, NodeConfig } from "@graphter/core";

const config:NodeConfig = {
  id: 'config',
  name: 'Graphter Config',
  description: 'The configuration that drives Graphter models',
  type: 'object',
  validation: [

  ],
  children: [
    {
      id: 'id',
      name: 'ID',
      description: 'The ID of the model',
      type: 'id',
      validation: [
        {
          type: 'json-schema',
          executeOn: AllValidationExecutionStages,
          options: {
            error: 'An ID is required',
            schema: {
              "minLength": 1
            },
          }
        },
        {
          type: 'json-schema',
          executeOn: AllValidationExecutionStages,
          options: {
            error: 'Only lowercase alphanumeric and hyphen characters are allowed in IDs',
            schema: {
              'type': 'string',
              'regexp': '/^[a-z0-9-]*$/',
            },
          }
        },
        {
          type: 'id-uniqueness',
          executeOn: AllValidationExecutionStages,
          options: {
            serviceId: 'config'
          }
        }
      ]
    },
    {
      id: 'name',
      name: 'Name',
      description: 'The model name',
      type: 'string',
      validation: [
        {
          type: 'json-schema',
          executeOn: AllValidationExecutionStages,
          options: {
            error: 'A name is required',
            schema: {
              "minLength": 1
            },
          }
        },
      ]
    },
    {
      id: 'description',
      name: 'Description',
      description: 'A description of the model',
      type: 'string',
      validation: [ ]
    },
    {
      id: 'type',
      name: 'Type',
      description: 'The type of data',
      type: 'data-select',
      children: [ ],
      validation: [ ],
      default: 'text',
      options: {
        service: 'node-renderer',
        keyPath: [ 'type' ],
        valuePath: [ 'name' ]
      }
    },
    {
      id: 'children',
      name: 'Children',
      type: 'conditional',
      validation: [],
      default: [],
      options: {
        siblingPath: ['type'],
        branches: [
          { condition: 'object', childId: 'object-children' },
          { condition: 'list', childId: 'list-children' }
        ]
      },
      children: [
        {
          id: 'object-children',
          name: 'Object properties',
          type: 'list',
          children: [
            {
              id: 'child',
              name: 'Object property',
              description: 'Like an authors name or the first line of an authors address',
              type: 'nested',
              options: {
                configId: 'config'
              }
            },
          ],
          validation: [ ],
          options: {
            itemSelectionBehaviour: 'CUSTOM'
          }
        },
        {
          id: 'list-children',
          name: 'List types',
          type: 'list',
          children: [
            {
              id: 'child',
              name: 'List type',
              description: 'Define what list items look like',
              type: 'nested',
              options: {
                configId: 'config'
              }
            },
          ],
          validation: [ ],
          options: {
            itemSelectionBehaviour: 'CUSTOM',
            maxItems: 1
          }
        }
      ]
    }
  ],
}

export default config