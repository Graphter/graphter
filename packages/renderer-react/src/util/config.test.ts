import { configUtils } from "./config";

describe('config', () => {
  describe('getAt', () => {

    it('should return the correct config when looking one step deep', () => {
      const result = configUtils.getAt({
        id: 'some-config-id',
        type: 'object',
        children: [ { id: 'another-config-id', type: 'string', } ]
      }, ['another-config-id'])
      expect(result.id).toBe('another-config-id')
    })

    it('should return the correct config when looking two steps deep', () => {
      const result = configUtils.getAt({
        id: 'some-config-id',
        type: 'object',
        children: [
          {
            id: 'another-config-id',
            type: 'list',
            children: [
              { id: 'and-another-config-id', type: 'string' }
            ]
          }
        ]
      }, ['another-config-id', 'and-another-config-id'])
      expect(result.id).toBe('and-another-config-id')
    })

    it('should return the supplied config straight out when an empty path is supplied', () => {
      const result = configUtils.getAt({
        id: 'some-config-id',
        type: 'string'
      }, [])
      expect(result.id).toBe('some-config-id')
    })

    it('should throw an error if no config is found', () => {
      const config = {
        id: 'some-config-id',
        type: 'object',
        children: [
          {
            id: 'another-config-id',
            type: 'list',
            children: [
              { id: 'and-another-config-id', type: 'string' }
            ]
          }
        ]
      }
      expect(() => configUtils.getAt(config, ['another-config-id', 'some-non-existent-config-id']))
        .toThrowErrorMatchingSnapshot()
    })

    it.each([null, undefined, {}, {id: 'some-invalid-config-id'}])
    ('should throw an error if invalid config %o is supplied', (invalidConfig) => {
      // @ts-ignore
      expect(() => configUtils.getAt(invalidConfig, [ 'some-invalid-config-id']))
        .toThrowErrorMatchingSnapshot()
    })

    it.each([[undefined], null, undefined])
    ('should throw an error if invalid path %o is supplied', (invalidPath) => {
      expect(() => configUtils.getAt({
        id: 'some-config-id',
        type: 'object',
        children: [ { id: 'some-child-config-id', type: 'string' } ]
        // @ts-ignore
      }, invalidPath))
        .toThrowErrorMatchingSnapshot()
    })

  })
})

export {}