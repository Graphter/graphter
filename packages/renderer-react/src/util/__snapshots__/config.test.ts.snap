// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config getAt should throw an error if invalid config { id: 'some-invalid-config-id' } is supplied 1`] = `"Could not find config at some-invalid-config-id within: {\\"id\\":\\"some-invalid-config-id\\"}."`;

exports[`config getAt should throw an error if invalid config {} is supplied 1`] = `"Could not find config at some-invalid-config-id within: {}."`;

exports[`config getAt should throw an error if invalid config null is supplied 1`] = `"Could not find config at some-invalid-config-id within: null."`;

exports[`config getAt should throw an error if invalid config undefined is supplied 1`] = `"Could not find config at some-invalid-config-id within: undefined."`;

exports[`config getAt should throw an error if invalid path [ undefined, [length]: 1 ] is supplied 1`] = `"Could not find config at  within: {\\"id\\":\\"some-config-id\\",\\"type\\":\\"object\\",\\"children\\":[{\\"id\\":\\"some-child-config-id\\",\\"type\\":\\"string\\"}]}."`;

exports[`config getAt should throw an error if invalid path null is supplied 1`] = `"Could not find config at undefined within: {\\"id\\":\\"some-config-id\\",\\"type\\":\\"object\\",\\"children\\":[{\\"id\\":\\"some-child-config-id\\",\\"type\\":\\"string\\"}]}."`;

exports[`config getAt should throw an error if invalid path undefined is supplied 1`] = `"Could not find config at undefined within: {\\"id\\":\\"some-config-id\\",\\"type\\":\\"object\\",\\"children\\":[{\\"id\\":\\"some-child-config-id\\",\\"type\\":\\"string\\"}]}."`;

exports[`config getAt should throw an error if no config is found 1`] = `"Could not find config at another-config-id/some-non-existent-config-id within: {\\"id\\":\\"some-config-id\\",\\"type\\":\\"object\\",\\"children\\":[{\\"id\\":\\"another-config-id\\",\\"type\\":\\"list\\",\\"children\\":[{\\"id\\":\\"and-another-config-id\\",\\"type\\":\\"string\\"}]}]}."`;
